File: shim-php/InventoryImport.php
<?php declare(strict_types=1);

final class UnsupportedMediaTypeException extends RuntimeException {}

function minorFromString(string $s): int {
    $s = str_replace(['$', ',', ' '], '', $s);
    if (!preg_match('/^-?\d+(?:\.\d+)?$/', $s)) {
        throw new InvalidArgumentException('invalid price');
    }
    $neg = str_starts_with($s, '-');
    if ($neg) $s = substr($s, 1);
    [$w, $f] = array_pad(explode('.', $s, 2), 2, '');
    $f = str_pad(substr($f, 0, 3), 3, '0');
    $c = (int)substr($f, 0, 2);
    if ((int)$f[2] >= 5) $c++;
    $v = ((int)$w) * 100 + $c;
    return $neg ? -$v : $v;
}

function normalizeHeading(string $s): string {
    return trim((string)preg_replace('/[^a-z0-9]+/', '_', strtolower(trim($s))), '_');
}

function parseCsvRows(string $raw): array {
    $first = strtok($raw, "\r\n") ?: '';
    $delim = substr_count($first, "\t") > substr_count($first, ',')
        ? "\t"
        : (substr_count($first, ';') > substr_count($first, ',') ? ';' : ',');
    $f = fopen('php://temp', 'r+');
    if ($f === false) throw new RuntimeException('temporary stream unavailable');
    fwrite($f, $raw);
    rewind($f);
    $head = fgetcsv($f, 0, $delim);
    if (!$head) { fclose($f); return []; }
    $head = array_map('normalizeHeading', $head);
    $rows = [];
    while (($r = fgetcsv($f, 0, $delim)) !== false) {
        if (!array_filter($r, fn($v) => trim((string)$v) !== '')) continue;
        $r = array_slice(array_pad($r, count($head), ''), 0, count($head));
        $rows[] = array_combine($head, $r);
    }
    fclose($f);
    return $rows;
}

function xlsxColumnIndex(string $ref): int {
    if (!preg_match('/^([A-Z]+)/i', $ref, $m)) return 0;
    $n = 0;
    foreach (str_split(strtoupper($m[1])) as $ch) $n = $n * 26 + (ord($ch) - 64);
    return max(0, $n - 1);
}

function xlsxXml(string $xml): SimpleXMLElement {
    $old = libxml_use_internal_errors(true);
    try {
        $node = simplexml_load_string($xml, SimpleXMLElement::class, LIBXML_NONET | LIBXML_COMPACT);
        if ($node === false) throw new InvalidArgumentException('invalid XLSX XML');
        return $node;
    } finally {
        libxml_clear_errors();
        libxml_use_internal_errors($old);
    }
}

/** @return array<int,array<int,string>> */
function parseXlsxMatrix(string $raw): array {
    if (!class_exists('ZipArchive') || !function_exists('simplexml_load_string')) {
        throw new UnsupportedMediaTypeException('XLSX requires PHP extensions zip and SimpleXML');
    }
    $tmp = tempnam(sys_get_temp_dir(), 'fvs-xlsx-');
    if ($tmp === false) throw new RuntimeException('cannot allocate XLSX temporary file');
    try {
        if (file_put_contents($tmp, $raw, LOCK_EX) === false) throw new RuntimeException('cannot stage XLSX');
        $zip = new ZipArchive();
        if ($zip->open($tmp) !== true) throw new InvalidArgumentException('invalid XLSX archive');
        try {
            $shared = [];
            $sharedRaw = $zip->getFromName('xl/sharedStrings.xml');
            if ($sharedRaw !== false) {
                $root = xlsxXml($sharedRaw);
                $ns = $root->getNamespaces(true);
                $d = isset($ns['']) ? $root->children($ns['']) : $root;
                foreach ($d->si as $si) {
                    $parts = $si->xpath('.//*[local-name()="t"]') ?: [];
                    $text = '';
                    foreach ($parts as $p) $text .= (string)$p;
                    $shared[] = $text;
                }
            }
            $sheetName = null;
            for ($i = 0; $i < $zip->numFiles; $i++) {
                $name = $zip->getNameIndex($i);
                if ($name !== false && preg_match('#^xl/worksheets/sheet\d+\.xml$#', $name)) {
                    $sheetName = $name; break;
                }
            }
            if ($sheetName === null) throw new InvalidArgumentException('xlsx has no worksheet');
            $sheetRaw = $zip->getFromName($sheetName);
            if ($sheetRaw === false) throw new InvalidArgumentException('cannot read XLSX worksheet');
            $root = xlsxXml($sheetRaw);
            $rowNodes = $root->xpath('//*[local-name()="row"]') ?: [];
            $rows = [];
            foreach ($rowNodes as $rowNode) {
                $vals = [];
                $cells = $rowNode->xpath('./*[local-name()="c"]') ?: [];
                foreach ($cells as $cell) {
                    $attrs = $cell->attributes();
                    $idx = xlsxColumnIndex((string)($attrs['r'] ?? 'A'));
                    $type = (string)($attrs['t'] ?? '');
                    $vNodes = $cell->xpath('./*[local-name()="v"]') ?: [];
                    $tNodes = $cell->xpath('.//*[local-name()="t"]') ?: [];
                    $text = $tNodes ? (string)$tNodes[0] : ($vNodes ? (string)$vNodes[0] : '');
                    if ($type === 's' && $text !== '' && ctype_digit($text)) $text = $shared[(int)$text] ?? '';
                    $vals[$idx] = $text;
                }
                if ($vals) {
                    $max = max(array_keys($vals));
                    $line = [];
                    for ($i = 0; $i <= $max; $i++) $line[] = $vals[$i] ?? '';
                    $rows[] = $line;
                }
            }
            return $rows;
        } finally {
            $zip->close();
        }
    } finally {
        @unlink($tmp);
    }
}

function matrixToAssocRows(array $matrix): array {
    if (!$matrix) return [];
    $head = array_map(fn($v) => normalizeHeading((string)$v), array_shift($matrix));
    $rows = [];
    foreach ($matrix as $r) {
        if (!array_filter($r, fn($v) => trim((string)$v) !== '')) continue;
        $r = array_slice(array_pad($r, count($head), ''), 0, count($head));
        $rows[] = array_combine($head, $r);
    }
    return $rows;
}

function normRow(array $r): array {
    foreach (['resort', 'unit_code', 'unit_name', 'check_in', 'check_out'] as $k) {
        if (trim((string)($r[$k] ?? '')) === '') throw new InvalidArgumentException('missing '.$k);
    }
    $src = trim((string)($r['source_ref'] ?? ''));
    if ($src === '') {
        $src = 'auto:'.substr(hash('sha256', implode('|', [$r['resort'], $r['unit_code'], $r['check_in'], $r['check_out']])), 0, 40);
    }
    $pm = trim((string)($r['price_minor'] ?? '')) !== ''
        ? (int)str_replace(',', '', (string)$r['price_minor'])
        : minorFromString((string)($r['price'] ?? ''));
    $active = !in_array(strtolower(trim((string)($r['active'] ?? '1'))), ['0', 'false', 'no', 'inactive'], true);
    $mode = strtolower(trim((string)($r['booking_mode'] ?? 'fixed')));
    if (!in_array($mode, ['fixed','floating'], true)) throw new InvalidArgumentException('invalid booking_mode');
    $floatGroup = trim((string)($r['float_group'] ?? ''));
    if ($mode === 'floating' && $floatGroup === '') throw new InvalidArgumentException('floating inventory requires float_group');
    $weekRaw = trim((string)($r['week_number'] ?? ''));
    $week = $weekRaw !== '' ? (int)$weekRaw : (int)(new DateTimeImmutable(substr(trim((string)$r['check_in']),0,10)))->format('W');
    if ($week < 1 || $week > 53) throw new InvalidArgumentException('invalid week_number');
    return [
        'source_ref' => $src,
        'resort' => trim((string)$r['resort']),
        'unit_code' => trim((string)$r['unit_code']),
        'unit_name' => trim((string)$r['unit_name']),
        'city' => trim((string)($r['city'] ?? '')),
        'country' => trim((string)($r['country'] ?? '')),
        'check_in' => substr(trim((string)$r['check_in']), 0, 10),
        'check_out' => substr(trim((string)$r['check_out']), 0, 10),
        'week_number' => $week,
        'booking_mode' => $mode,
        'float_group' => $floatGroup,
        'max_guests' => (int)($r['max_guests'] ?? 2),
        'price_minor' => $pm,
        'currency' => strtoupper(trim((string)($r['currency'] ?? 'MXN'))),
        'image_url' => trim((string)($r['image_url'] ?? '')),
        'short_description' => trim((string)($r['short_description'] ?? '')),
        'season' => trim((string)($r['season'] ?? '')),
        'latitude' => trim((string)($r['latitude'] ?? '')),
        'longitude' => trim((string)($r['longitude'] ?? '')),
        'property_type' => trim((string)($r['property_type'] ?? '')),
        'bedrooms' => (int)($r['bedrooms'] ?? 0),
        'bathrooms' => (float)($r['bathrooms'] ?? 0),
        'rating' => (float)($r['rating'] ?? 0),
        'amenities' => array_values(array_filter(array_map('trim', preg_split('/[,;|]/', (string)($r['amenities'] ?? '')) ?: []))),
        'active' => $active,
    ];
}

function importInventoryBytes(FvsCore $core, string $raw, string $filename): array {
    $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
    if (in_array($ext, ['csv', 'tsv', 'txt'], true)) {
        $rows = parseCsvRows($raw);
    } elseif (in_array($ext, ['xlsx', 'xlsm'], true)) {
        $rows = matrixToAssocRows(parseXlsxMatrix($raw));
    } else {
        throw new UnsupportedMediaTypeException('Supported inventory files: CSV, TSV, XLSX, XLSM');
    }
    $rep = ['rows_seen' => count($rows), 'inserted' => 0, 'updated' => 0, 'rejected' => 0, 'errors' => []];
    foreach ($rows as $i => $r) {
        try {
            $created = $core->inventoryUpsert(normRow($r));
            $rep[$created ? 'inserted' : 'updated']++;
        } catch (Throwable $x) {
            $rep['rejected']++;
            if (count($rep['errors']) < 50) $rep['errors'][] = ['row' => $i + 2, 'error' => substr($x->getMessage(), 0, 240)];
        }
    }
    return $rep;
}

File: shim-python/inventory_import.py
from __future__ import annotations
import csv, hashlib, io, re, zipfile
from datetime import date
from decimal import Decimal, ROUND_HALF_UP
from xml.etree import ElementTree as ET
import core

REQ={"resort","unit_code","unit_name","check_in","check_out"}

def _col(ref):
    m=re.match(r"([A-Z]+)",ref or "A");n=0
    for ch in (m.group(1) if m else "A"):n=n*26+ord(ch)-64
    return n-1

def _xlsx_rows(raw:bytes):
    with zipfile.ZipFile(io.BytesIO(raw)) as z:
        shared=[]
        if "xl/sharedStrings.xml" in z.namelist():
            root=ET.fromstring(z.read("xl/sharedStrings.xml"))
            for si in root:shared.append("".join(t.text or "" for t in si.iter() if t.tag.endswith("}t")))
        sheet=next((n for n in z.namelist() if n.startswith("xl/worksheets/sheet") and n.endswith(".xml")),None)
        if not sheet:raise ValueError("xlsx has no worksheet")
        root=ET.fromstring(z.read(sheet));rows=[]
        for r in (x for x in root.iter() if x.tag.endswith("}row")):
            vals={}
            for c in (x for x in r if x.tag.endswith("}c")):
                idx=_col(c.attrib.get("r","A"));typ=c.attrib.get("t");v=next((x for x in c if x.tag.endswith("}v")),None);inline=next((x for x in c.iter() if x.tag.endswith("}t")),None)
                text=inline.text if inline is not None else (v.text if v is not None else "")
                if typ=="s" and text: text=shared[int(text)]
                vals[idx]=text or ""
            if vals:rows.append([vals.get(i,"") for i in range(max(vals)+1)])
        return rows

def rows_from_bytes(raw:bytes,filename:str):
    ext=filename.lower().rsplit('.',1)[-1] if '.' in filename else 'csv'
    if ext in {'xlsx','xlsm'}: rows=_xlsx_rows(raw)
    else:
        text=raw.decode('utf-8-sig');dialect=csv.Sniffer().sniff(text[:4096],delimiters=',;\t') if text.strip() else csv.excel
        rows=list(csv.reader(io.StringIO(text),dialect))
    if not rows:return []
    headers=[re.sub(r'[^a-z0-9]+','_',str(x).strip().lower()).strip('_') for x in rows[0]]
    return [dict(zip(headers,r)) for r in rows[1:] if any(str(x).strip() for x in r)]

def _minor(row):
    if str(row.get('price_minor','')).strip():return int(str(row['price_minor']).replace(',','').strip())
    raw=str(row.get('price','')).replace('$','').replace(',','').strip()
    return int((Decimal(raw)*100).quantize(Decimal('1'),rounding=ROUND_HALF_UP))

def normalize(row):
    missing=[k for k in REQ if not str(row.get(k,'')).strip()]
    if missing:raise ValueError('missing '+','.join(sorted(missing)))
    src=str(row.get('source_ref','')).strip()
    if not src:
        basis='|'.join(str(row.get(k,'')) for k in ('resort','unit_code','check_in','check_out'))
        src='auto:'+hashlib.sha256(basis.encode()).hexdigest()[:40]
    active=str(row.get('active','1')).strip().lower() not in {'0','false','no','inactive'}
    ci=str(row['check_in']).strip()[:10]
    mode=str(row.get('booking_mode') or 'fixed').strip().lower()
    if mode not in {'fixed','floating'}: raise ValueError('invalid booking_mode')
    fg=str(row.get('float_group','')).strip()
    if mode=='floating' and not fg: raise ValueError('floating inventory requires float_group')
    week_raw=str(row.get('week_number','')).strip()
    week=int(week_raw) if week_raw else date.fromisoformat(ci).isocalendar().week
    amenities=[x.strip() for x in re.split(r'[,;|]',str(row.get('amenities',''))) if x.strip()]
    return {'source_ref':src,'resort':str(row['resort']).strip(),'unit_code':str(row['unit_code']).strip(),'unit_name':str(row['unit_name']).strip(),'city':str(row.get('city','')).strip(),'country':str(row.get('country','')).strip(),'check_in':ci,'check_out':str(row['check_out']).strip()[:10],'week_number':week,'booking_mode':mode,'float_group':fg,'max_guests':int(row.get('max_guests') or 2),'price_minor':_minor(row),'currency':str(row.get('currency') or 'MXN').strip().upper(),'image_url':str(row.get('image_url','')).strip(),'short_description':str(row.get('short_description','')).strip(),'season':str(row.get('season','')).strip(),'latitude':str(row.get('latitude','')).strip(),'longitude':str(row.get('longitude','')).strip(),'property_type':str(row.get('property_type','')).strip(),'bedrooms':int(row.get('bedrooms') or 0),'bathrooms':float(row.get('bathrooms') or 0),'rating':float(row.get('rating') or 0),'amenities':amenities,'active':active}

def import_bytes(raw:bytes,filename:str):
    rows=rows_from_bytes(raw,filename);report={'rows_seen':len(rows),'inserted':0,'updated':0,'rejected':0,'errors':[]}
    for i,row in enumerate(rows,2):
        try:
            created=core.inventory_upsert(normalize(row));report['inserted' if created else 'updated']+=1
        except Exception as e:
            report['rejected']+=1
            if len(report['errors'])<50:report['errors'].append({'row':i,'error':str(e)[:240]})
    return report
